kernel: clarify and assert interrupts aren't disabled upon context switch

mathieuchopstm opened the pull request:
The `do_swap()` routine used when `CONFIG_USE_SWITCH=y` asserts that caller thread does not hold any spinlock when `CONFIG_SPIN_VALIDATE` is enabled. However, there is no similar check in place when `CONFIG_USE_SWITCH=n`.

Copy this assertion in the `USE_SWITCH=n` implementation of `z_swap_irqlock()`.

mathieuchopstm committed:
kernel: assert no spinlock is held on swap when !USE_SWITCH

The do_swap() routine used when CONFIG_USE_SWITCH=y asserts that caller
thread does not hold any spinlock when CONFIG_SPIN_VALIDATE is enabled.
However, there is no similar check in place when CONFIG_USE_SWITCH=n.

Copy this assertion in the USE_SWITCH=n implementation of z_swap_irqlock().

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

andyross reviewed (dismissed):
For clarity: this is useful for architectures like arm32 that have an arch_swap() and miss the check elsewhere.  The error being detected is always wrong.

krish2718 reviewed (dismissed):
Thanks, this caught the bug in my case where a thread holding the spinlock tries to sleep (mutex).

andyross commented:
My vote would just be to remove that section of the docs and rip the bandaid off with the warning (which can always be disabled via kconfig anyway).   There are precisely zero circumstances where breaking a critical section (!) lock with a context switch (!!) cannot lead to tears.

Code that does that is not only breaking its own carefully curated lock in an immensely clever way that it clearly thought carefully about and knows has no unexpected interactions.  Because of the recursive behavior of irq_lock() it's also breaking the locks taken by **all the callers up the stack**, who had no idea that calling into this obnoxiously clever subsystem was a synchronization trap.

It just can't work.  We never should have documented it (and I didn't even know we had!).

krish2718 commented:
> It is a fatal error to hold the interrupt lock upon context switch

How about the infamous Linux wording `BUG: Scheduling while atomic`?

mathieuchopstm commented:
> > It is a fatal error to hold the interrupt lock upon context switch
> 
> How about the infamous Linux wording `BUG: Scheduling while atomic`?

We already have a runtime message :slightly_smiling_face: 
https://github.com/zephyrproject-rtos/zephyr/blob/47b07e5a09efdec59247770b5563b75b4ff75324/kernel/include/kswap.h#L98-L100

What needs updating is the documentation for `irq_lock()` as it isn't consistent with the rest of the kernel.

bjarki-andreasen commented:
I think we need the explicit wording of _"It is a fatal error to hold the interrupt lock upon context switch"_, since a thread can force a reschedule/context switch at any time by calling k_yield() for example.

mathieuchopstm commented:
> I think we need the explicit wording of _"It is a fatal error to hold the interrupt lock upon context switch"_, since a thread can force a reschedule/context switch at any time by calling k_yield() for example.

I don't think `k_yield()` is `isr-ok` though?

(Really, my `Only isr-ok functions may be called while holding the interrupt lock` is a non-kernel-wording way of saying `It is a fatal error to hold the interrupt lock upon context switch` - I don't mind adding it anyways if you feel like it's better though)

mathieuchopstm committed:
kernel: irq: update irq_lock() documentation regarding context switches

Upon context switch, the virtual "interrupt lock" acquired by irq_lock()
must not be "held". However, the current documentation for irq_lock() says
that it is perfectly valid to hold it (!), and that a suspended thread will
hold the "interrupt lock" upon being scheduled again (!!).

Update the documentation to remove the outdated section and indicate that
context switching while holding the interrupt lock is not allowed.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
kernel: spinlock: update k_spin_lock() documentation wrt context switch

Threads must not attempt to context switch if they are holding a spinlock.

Add this information to the documentation for k_spin_lock().

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

andyross reviewed (approved):
Looks clear to me.

bjarki-andreasen reviewed (approved):
Seems perfect
